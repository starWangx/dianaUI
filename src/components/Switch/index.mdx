---
name: Switch —— 开关
route: /switch
---
import { Playground, Props } from 'docz'
import Switch from './index.tsx'

# Switch 开关组件

<Props of={Switch} />

## Demo
<Playground>
  <Switch />
  <br />
  <br />
  <Switch
    className='custom'
    renderChecked='开'
    renderUnChecked='关'
    defaultChecked
    autoFocus
  />
  <br />
  <br />
  <Switch
    disabled
  />
</Playground>

## API - Props
- `style` 样式style属性
- `className` 类名
- `autoFocus` 自动聚焦
- `checked` 是否 `checked`，如果传入该值，则 `Switch` 组件为受控组件
- `defaultChecked` 状态`checked` 的初始值
- `renderChecked` 状态是 `checked` 时展示的内容
- `renderUnChecked` 状态是 `unchecked` 时展示的内容
- `onChange` 回调 - 当`checked` 状态更改时
- `onClick` 回调 - 点击
- `onMouseUp` 回调 - 鼠标抬起

