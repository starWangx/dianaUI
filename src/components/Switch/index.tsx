import React, { SFC } from 'react';
import { Switch, SwitchProps } from '@hupu/rc-basic';
import styled, { keyframes } from 'styled-components';
import * as C from '../../styles/theme';

export interface SwitchUIProps extends SwitchProps { };

const switchOn = keyframes`
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.25);
  }
  100% {
    transform: scale(1.1);
  }
`;

const switchOff = keyframes`
  0% {
    transform: scale(1.1);
  }
  100% {
    transform: scale(1);
  }
`;

const StyledSwitch = styled(Switch)`
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  width: 44px;
  height: 22px;
  line-height: 20px;
  padding: 0;
  vertical-align: middle;
  border-radius: 20px 20px;
  border: 1px solid #ccc;
  background-color: #ccc;
  cursor: pointer;
  transition: all .3s cubic-bezier(0.35, 0, 0.25, 1);

  :after{
    position: absolute;
    width: 18px;
    height: 18px;
    left: 2px;
    top: 1px;
    border-radius: 50% 50%;
    background-color: #fff;
    content: " ";
    cursor: pointer;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.26);
    transform: scale(1);
    transition: left .3s cubic-bezier(0.35, 0, 0.25, 1);
    animation-timing-function: cubic-bezier(0.35, 0, 0.25, 1);
    animation-duration: .3s;
    animation-name: ${switchOff};
  }

  :hover:after {
    transform: scale(1.1);
    animation-name: ${switchOn};
  }

  :focus {
    box-shadow: 0 0 0 2px fade(${C.colorBlue1890ff}, 20%);
    outline: none;
  }

  :focus:hover {
    box-shadow: none;
  }

  & .diana-switch-inner {
    color: #fff;
    font-size: 12px;
    position: absolute;
    left: 24px;
    top: 0;
  }

  &.diana-switch-checked {
    border: 1px solid ${C.colorBlue1A68FF};
    background-color: ${C.colorBlue1A68FF};
    :after{
      left: 22px;
    }
    .diana-switch-inner {
      left: 6px;
    }
  }

  &.diana-switch-disabled {
    cursor: no-drop;
    background: #ccc;
    border-color: #ccc;
    :after{
      background: #9e9e9e;
      animation-name: none;
      cursor: no-drop;
    }
    :hover:after{
      transform: scale(1);
      animation-name: none;
    }
  }
`;

const SwitchUI: SFC<SwitchUIProps> = props => {
  return (
    <StyledSwitch
      { ...props }
      prefixCls='diana-switch'
    />
  );
};

export default SwitchUI;