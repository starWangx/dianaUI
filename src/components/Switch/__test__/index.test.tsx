import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Switch from '../index';

configure({ adapter: new Adapter() });

describe('Switch', () => {
  it('renders correctly - shallow', () => {
    const wrapper = shallow(<Switch />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - render', () => {
    const wrapper = render(<Switch />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - mount', () => {
    const wrapper = mount(<Switch />);

    expect(wrapper).toMatchSnapshot();
  });
});
