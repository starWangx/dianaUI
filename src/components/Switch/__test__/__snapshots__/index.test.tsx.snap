// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders correctly - mount 1`] = `ReactWrapper {}`;

exports[`Switch renders correctly - render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "aria-checked": "false",
      "class": "diana-switch sc-bdVaJa bHxQXe",
      "role": "switch",
      "type": "button",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "diana-switch-inner",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-checked": undefined,
      "class": undefined,
      "role": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-checked": undefined,
      "class": undefined,
      "role": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Switch renders correctly - shallow 1`] = `ShallowWrapper {}`;
